### ISSUE TITLE
Sometimes items in unimported files don't show in code completion (due to completion time budget)

### ISSUE THREAD

**Issue by lukehutch** [Dart-Code/Dart-Code#4760]
**Describe the bug**

I had a class called `AvatarBadge` that existed in the file `profile_photo_widgets.dart`:

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/2d03cc56-c947-43ce-8d8d-0e26a4097a55)

I removed the class from that file, and pasted it into a new file in the same folder, `avatar_badge.dart`:

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/77db24b3-1bde-48e8-999c-44ca30b432b8)

Now I have some broken references to `AvatarBadge` back in `profile_photo_widgets.dart`:

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/84092d3b-a7f8-42b3-bd69-146a3641a15f)

Pressing Ctrl+Space after the missing symbol says "No suggestions".

"Ctrl + P -> Reload window" does not fix the problem.

**Please complete the following information:**
 - Operating System and version: Linux
 - VS Code version: latest
 - Dart extension version: latest
 - Dart/Flutter SDK version: latest
 - Target device (if the issue relates to Flutter debugging):

**Comment by lukehutch** [comment]
Pressing Ctrl+Space multiple times seems to eventually cause the file to be found and indexed. Sometimes it happens a minute later though, so I suspect something else is triggering the analyzer to find that file.

**Comment by DanTup** [comment]
@lukehutch can you reproduce this? Do you know whether the server was "Analyzing..." while this was happening? Pressing Ctrl+Space a lot _shouldn't_ really change anything, but if the server was busy analyzing it's possible the results would change because of that.

**Comment by lukehutch** [comment]
Yes, it happens all the time. Sometimes I have had to quit Code and reopen just to get the analyzer to pick up the new files.

**Comment by DanTup** [comment]
@lukehutch can you reproduce it in a new test project with the [instrumentation log](https://dartcode.org/docs/logging/#analyzer-instrumentation) enabled?

**Comment by lukehutch** [comment]
I suspect I won't be able to reproduce this in a new (small) project, otherwise I'm sure other people would have run into this a long time ago. It's probably something specific about my setup.

If I enable the instrumentation log on my (large) project, I'm assuming a large chunk of the program structure of my project will be generated in the captured instrumentation logs? If so, I wouldn't want to post that here, but I could email you a link to a Drive file or something.

**Comment by lukehutch** [comment]
@DanTup I suspect this may actually be an issue with the indexer simply taking a long time to index all the dependencies of my project (since sometimes, after 10-30 seconds, the Ctrl+Space menu does show the symbol I was trying to complete).

Would it be possible to simply add an info line at the top of that Ctrl+Space menu saying "Indexing source..." or something, whenever the indexer/analyzer is running?

**Comment by DanTup** [comment]
> Would it be possible to simply add an info line at the top of that Ctrl+Space menu saying "Indexing source..." or something, whenever the indexer/analyzer is running?

This should actually be part of normal analysis, and there should be an "analyzing..." indicator.


> If I enable the instrumentation log on my (large) project, I'm assuming a large chunk of the program structure of my project will be generated in the captured instrumentation logs? If so, I wouldn't want to post that here, but I could email you a link to a Drive file or something.

You're right, the instrumentation file will contain parts of your source code (when you open a file, the whole file will show up in the log). You're welcome to share logs privately (you can email them to me at logs@dartcode.org), but you should still be careful not to send anything sensitive (I don't want private source code or passwords 🙃). If you capture a log, I could also provide some notes on what to look for so you might be able to extract the useful parts.

Out of interest, how big is the workspace? The output of `code --status` when you have the project open might be helpful.

**Comment by lukehutch** [comment]
The main Flutter project is probably 100 Dart source files (which is not enormous, but it's not small). I have about 60 pub.dev deps.

```console
$ code --status
Version:          Code 1.82.2 (abd2f3db4bdb28f9e95536dfa84d8479f1eb312d, 2023-09-14T05:51:20.981Z)
OS Version:       Linux x64 6.4.12-200.fc38.x86_64
CPUs:             AMD Ryzen 9 3950X 16-Core Processor (32 x 2120)
Memory (System):  125.70GB (85.93GB free)
Load (avg):       1, 1, 1
VM:               0%
Screen Reader:    no
Process Argv:     click.code-workspace --crash-reporter-id 11879ce0-438e-429c-8d0c-4da9fd848f93
GPU Status:       2d_canvas:                              enabled
                  canvas_oop_rasterization:               disabled_off
                  direct_rendering_display_compositor:    disabled_off_ok
                  gpu_compositing:                        enabled
                  multiple_raster_threads:                enabled_on
                  opengl:                                 enabled_on
                  rasterization:                          enabled
                  raw_draw:                               disabled_off_ok
                  video_decode:                           enabled
                  video_encode:                           disabled_software
                  vulkan:                                 disabled_off
                  webgl:                                  enabled
                  webgl2:                                 enabled
                  webgpu:                                 disabled_off

CPU %	Mem MB	   PID	Process
    0	   257	596552	code main
    0	     0	596559	   zygote
    0	   257	596593	     gpu-process
    0	     0	596628	       broker
    0	     0	596560	   zygote
    0	     0	596562	     zygote
    0	   386	596616	window [1] (image.dart - click (Workspace) - Visual Studio Code)
    0	   257	1237053	window [4] (create_test_profiles.dart - click_server - Visual Studio Code)
    0	     0	596598	   utility-network-service
    0	     0	596676	shared-process
    0	     0	658133	ptyHost
    0	     0	3102841	     /usr/bin/bash --init-file /usr/share/code/resources/app/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh
    0	   257	1683705	extensionHost [4]
    0	   901	1683738	     /opt/flutter/bin/cache/dart-sdk/bin/dart language-server --protocol=lsp --client-id=VS-Code --client-version=3.72.2
    0	     0	1822426	     /opt/flutter/bin/cache/dart-sdk/bin/dart debug_adapter
    0	   515	1822435	       electron-nodejs (vm.js )
    0	     0	1683710	fileWatcher [4]
    0	   515	3102689	extensionHost [1]
    0	   257	3103202	     /usr/java/latest/bin/java -Dfile.encoding=utf8 -jar /home/luke/.vscode/extensions/vscjava.vscode-gradle-3.12.7/lib/gradle-language-server.jar
    0	     0	3103307	     /usr/java/latest/bin/java -Dfile.encoding=UTF-8 -classpath /home/luke/.vscode/extensions/vscjava.vscode-gradle-3.12.7/lib/gradle-server.jar com.github.badsyntax.gradle.GradleServer 38087
    0	   386	3103428	     /home/luke/.vscode/extensions/redhat.java-1.22.1-linux-x64/jre/17.0.8.1-linux-x86_64/bin/java --add-modules=ALL-SYSTEM --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/sun.nio.fs=ALL-UNNAMED -Declipse.application=org.eclipse.jdt.ls.core.id1 -Dosgi.bundles.defaultStartLevel=4 -Declipse.product=org.eclipse.jdt.ls.core.product -Djava.import.generatesMetadataFilesAtProjectRoot=false -Dfile.encoding=utf8 -XX:+UseParallelGC -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Xmx1G -Xms100m -Xlog:disable -javaagent:/home/luke/.vscode/extensions/redhat.java-1.22.1-linux-x64/lombok/lombok-1.18.28.jar -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/home/luke/.config/Code/User/workspaceStorage/9a8af0226dd93409e27ac99b624ed091/redhat.java -Daether.dependencyCollector.impl=bf -jar /home/luke/.vscode/extensions/redhat.java-1.22.1-linux-x64/server/plugins/org.eclipse.equinox.launcher_1.6.500.v20230717-2134.jar -configuration /home/luke/.config/Code/User/globalStorage/redhat.java/1.22.1/config_linux -data /home/luke/.config/Code/User/workspaceStorage/9a8af0226dd93409e27ac99b624ed091/redhat.java/jdt_ws
    0	     0	3103476	     /home/luke/.vscode/extensions/ms-vscode.cpptools-1.17.5-linux-x64/bin/cpptools
    0	     0	3811071	     electron-nodejs (/usr/share/code/code --ms-enable-electron-run-as-node /usr/share/code/resources/app/extensions/markdown-language-features/server/dist/node/workerMain --node-ipc --clientProcessId=3102689)
    0	     0	3820666	     electron-nodejs (/usr/share/code/code --ms-enable-electron-run-as-node /usr/share/code/resources/app/extensions/json-language-features/server/dist/node/jsonServerMain --node-ipc --clientProcessId=3102689)
    0	     0	3955193	     electron-nodejs (config.js )
    0	  1416	3955208	     /opt/flutter/bin/cache/dart-sdk/bin/dart language-server --protocol=lsp --client-id=VS-Code --client-version=3.72.2
    0	     0	3955219	     /opt/flutter/bin/cache/dart-sdk/bin/dart devtools --machine --try-ports 10 --allow-embedding --port 9104
    0	   129	3102694	fileWatcher [1]

Workspace Stats: 
|  Window (create_test_profiles.dart - click_server - Visual Studio Code)
|  Window (image.dart - click (Workspace) - Visual Studio Code)
|    Folder (click_client): 36 files
|      File types: dart(21) md(2) yaml(2) json(1) gitignore(1) lock(1)
|      Conf files:
|    Folder (click_server): 200 files
|      File types: jpg(50) dart(35) yaml(27) tf(13) json(10) sql(6) sh(4)
|                  gitignore(2) md(2) svg(2)
|      Conf files: launch.json(1) dockerfile(1)
|      Launch Configs: dart(2)
|    Folder (click_flutter): more than 20366 files
|      File types: xml(4017) flat(3727) json(1671) png(1330) len(375) jar(333)
|                  class(246) bin(199) dex(195) txt(168)
|      Conf files: launch.json(1) settings.json(1)
|      Launch Configs: dart(3)
```

**Comment by lukehutch** [comment]
@DanTup Here's a screen recording I just captured...

I added the dependency `serverpod_chat_flutter`, which defines the class `ChatController`. VS Code ran `pub get` when it noticed the yaml file changing, but Ctrl+Space over `ChatController` failed to find the new library. I tried reloading the window, and autocomplete still could not find the new dep. I ran `flutter pub get` from the commandline, and nothing changed. I started the screen recording, and looked to see that the analyzer wasn't still running (it wasn't). I tried Ctrl+Space after ChatController again, and it still couldn't find the symbol. Then I tried it a second time right after, on a whim, and now suddenly the symbol was found.

https://github.com/Dart-Code/Dart-Code/assets/811305/ec9f51b7-cccb-4867-885b-5f9d82866944

This pattern is very common -- symbols aren't found when new deps are added, or even sometimes when I define new classes within my project. However, the period of time between when I start trying to auto-import a class via Ctrl+Space and when the class actually starts popping up in the autocomplete box varies between seconds and minutes.

**Comment by DanTup** [comment]
Ok, I think I might understand what's happening here. From your examples, I think these are symbols that are _not yet imported_ into the current file?

Because the set of not-yet-imported symbols can be very large (for example can include all pub packages you depend on), there is a time budget on it. If compiling the list is not complete when this budget runs out, then we just send the completions that have already been computed. If we didn't do this, code completion could seem slow in large workspaces even though in the majority of cases you aren't typing unimported symbols. Although LSP has support for it, VS Code does not currently support "streaming" code completions - so we have to trade off being complete versus being fast.

Generally, the budget should not be hit - but the first few times you invoke completion in a session (or after adding new dependencies, etc.), or if your machine just happens to be slightly busy/slower you might hit it.

If you're seeing this a lot in this workspace you could try increasing the budget to see if the issue goes away:

```js
// default value is 100ms
"dart.completionBudgetMilliseconds": 500,
```

VS Code might tell you this setting is not valid - you can just ignore that. Generally we don't recommend users modifying this value except for understanding issues like this. If you are frequently hitting the budget, we might want to collect some more info about your workspace and see if there are optimisations that can be made.

**Comment by lukehutch** [comment]
> ```js
> // default value is 100ms
> "dart.completionBudgetMilliseconds": 500,
> ```

I get `Unknown Configuration Setting` for this, in the `settings.json` workspace settings.

**Comment by DanTup** [comment]
> This sounds like an indexing optimization problem? The time/space tradeoff says (in this case) that you can fit into any time budget, if you're willing to pay the space cost for the index.

I think it's a little more nuanced than that - there's time to build and maintain an index too. You frequently invoke code completion while you're typing, so you may be invalidating the current file (and potentially files that import it) and the completion request will be blocked behind updating the index. How long that takes depends on many things (and ofc there are trade-offs with how granular the invalidations are).

If this is a common (or severe) issue, we should do some profiling and try to improve the performance. It's not something that comes up much (at least that I know of) and is usually restricted to the first few invocations. It sounds like you might be seeing it more than that though - it would be interesting to see what timings show up in the diagnostic pages once you've increased the budget (so the requests aren't being stopped early).

You can access the diagnostic pages by running the **Dart: Open Analyzer Diagnostics** from the command palette (`F1`) then clicking **Completion** in the nav to get the timings.

If you can repro this issue on a public project (so I can test with a project of the same size/same dependencies), that would  be useful too - as I could test it here and do some CPU profiling to understand where the time goes in the cases where this happens.

**Comment by DanTup** [comment]
@lukehutch were you able to repro this on a public project I can test with? Or able to test it with a high budget and see what sort of times you're seeing if they're not cut short?

**Comment by lukehutch** [comment]
@DanTup I raised the budget as you suggested, and I don't *think* I have seen this problem since, but honestly I haven't been paying close enough attention, and it was so commonplace before that it might have happened a few times without me thinking of this issue report.

If this mostly or completely fixed the problem, then this raises another issue: the budget of 500ms that you suggested is a lot less than the time over which autocomplete of symbols in other files is sometimes broken (many seconds, or even minutes). This makes me think that if autocomplete fails once, then it doesn't get another chance once the analyzer's current run on the new files is complete, until the analyzer is run again on those new files. Otherwise I would have to increase the budget to several minutes to get the fixed behavior.

I haven't tried to reproduce this with other projects, but in general, you have to do some sort of major refactoring to trigger this -- e.g. rename a class, and also rename the file that the class is in, then you go to another class that refers to that changed class, and observe that autocomplete on the new class name no longer works. If you manually add an import to the new filename at the top of the referring class, that usually triggers the analyzer to start analyzing the renamed class, which fixes the problem immediately.

**Comment by DanTup** [comment]
> If this mostly or completely fixed the problem, then this raises another issue: the budget of 500ms that you suggested is a lot less than the time over which autocomplete of symbols in other files is sometimes broken (many seconds, or even minutes).

Right, the timer you're changing here is the budget allowed _per completion_. It's possible that the many seconds/few minutes you're seeing the issue is because some things have been invalidated and the server is running slightly slower (or your machine was just busier with other things). It's also the case that completion gets faster over the first few calls (probably because of JIT optimisation - though it's not something I'm entirely familiar with).

But to be clear - increasing the budget is _not_ a fix. The suggestion to do this was that I wanted to see some timings with the budget raised, so we can see how long the completion requests are taking if they are not cut short. We need to fix/optimize this case, but it's unfortunately more difficult to do so without more information (such as timings, repros or more detailed info about the size of the workspace).

Anything you can do to either get me closer to reproducing this, or timings from the diagnostic pages (instructions above) with the budget raised would be appreciated.

Thanks!

**Comment by lukehutch** [comment]
I'll keep an eye on this until I have something more detailed -- thanks, Dan!
